core/src/main/java/com/scalar/db/util/groupcommit/GroupCommitter.java

```diff
@@ -0,0 +1,190 @@
+package com.scalar.db.util.groupcommit;
+
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import com.google.common.util.concurrent.Uninterruptibles;
+import com.scalar.db.util.groupcommit.KeyManipulator.Keys;
+import java.io.Closeable;
+import java.time.Instant;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A group committer which is responsible for the group and slot managements and emits ready groups.
+ * This class receives 4 generic types as keys. But it's likely only one or a few key types are
+ * enough. The reason to use the 4 keys is to detect wrong key usages at compile time.
+ *
+ * @param <PARENT_KEY> A key type to NormalGroup which contains multiple slots and is
+ *     group-committed.
+ * @param <CHILD_KEY> A key type to slot in NormalGroup which can contain a value ready to commit.
+ * @param <FULL_KEY> A key type to DelayedGroup which contains a single slot and is
+ *     singly-committed.
+ * @param <EMIT_KEY> A key type that Emitter can interpret.
+ * @param <V> A value type to be set to a slot.
+ */
+public class GroupCommitter<PARENT_KEY, CHILD_KEY, FULL_KEY, EMIT_KEY, V> implements Closeable {
```

Comment:
I think we can put `@ThreadSafe` or `@NotThreadSafe` for each class.

Reply:
Fixed in https://github.com/scalar-labs/scalardb/pull/1607/commits/e58191815c85f3d27ed4a4c7e8aa8d36549b9976